core/src/main/java/org/jdbi/v3/core/result/RowReducer.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.result;
+
+import java.util.stream.Stream;
+
+/**
+ * Reduces row data from a {@link java.sql.ResultSet} into a stream of result
+ * elements. Useful for collapsing one-to-many joins.
+ *
+ * @param <C> mutable result container type
+ * @param <R> result element type
+ * @see ResultBearing#reduceRows(RowReducer)
+ */
+public interface RowReducer<C, R> {
+  /**
+   * Returns a new, empty result container.
+   *
+   * @return a new result container.
+   */
+  C container();
+
+  /**
+   * Accumulate data from the current row into the result container. Do not attempt
+   * to accumulate the {@link RowView} itself into the result container--it is only
+   * valid within the {@link RowReducer#accumulate(Object, RowView) accumulate()}
+   * method invocation. Instead, extract mapped types from the RowView by calling
+   * {@code RowView.getRow()} or {@code RowView.getColumn()} and store those values
+   * in the container.
+   *
+   * @param container the result container
+   * @param rowView row view over the current result set row.
+   */
+  void accumulate(C container, RowView rowView);
+
+  /**
+   * Returns a stream of result elements from the result container.
+   *
+   * @param container the result container
+   * @return stream of result elements.
+   */
+  Stream<R> stream(C container);
```

Comment:
Is returning a Stream a good API here, rather than just using the container?  (which the caller may then stream, or just use as-is)

There is no possibility to actually stream the results here, due to the context closing before the stream is presented to the caller.

Reply:
I chose stream here specifically so we could take advantage of the existing `JdbiCollectors` registry to collect the reduced elements into any supported container.

If the reducer returns the final container (e.g. `List<T>`, `Optional<T>`) instead of an intermediate stream of elements, then you have to write a reduce for every container type.

This way you can use the same reducer for all container types, provided the container is supported / registered into `JdbiCollectors`.

Reply:
I considered pushing this down into sqlobject, but couldn't make the generics work in the annotation:

```java
interface RowReducer<C, R> {
  C container();
  void accumulate(C, RowView);
  R result(C container);
}

@interface UseRowReducer {
  Class<? extends RowReducer<?, Stream<?>>> value();
}

class MyRowReducer implements RowReducer<Anything, Stream<Result>> { ... }

interface MyDao {
  @SqlQuery("...")
  @UseRowReducer(MyRowReducer.class)
  void doTheThing();
}
```

This caused a compiler error on `@UseRowReducer`, because `RowReducer<Anything,Stream<Result>` does not extend `RowReducer<?, Stream<?>>`. Generic wildcards just break down when you try to use nested parameters.

Reply:
You might need to write `Class<? extends RowReducer<?, ? extends Stream<?>>>` or something like that.

Reply:
> If the reducer returns the final container (e.g. List<T>, Optional<T>) instead of an intermediate stream of elements, then you have to write a reducer for every container type.

This is the primary reason I prefer the reducer returning a stream. The generics issues in SQL Object are secondary.

Reply:
So I've confirmed that the annotation would work as you suggested using `? extends Stream<?>`, however I still think we're better off with `RowReducer` producing a stream.

In my first prototype, the reducer returned the full container object instead of the stream:

```java
interface RowReducer<C,R> {
  C container();
  void accumulate(C container, RowView rowView);
  R result(C container);
}
```

Unfortunately, this meant that if I wanted to reduce into different containers (e.g. `List` vs `Optional`), I had to implement essentially the same reducer repeatedly, once for each container type:

```java
class ListOfContactPhoneReducer
    implements RowReducer<Map<Long, Contact>, List<Contact>> {
  Map<Long, Contact> container() {
    return new LinkedHashMap<>();
  }

  void accumulate(Map<Long, Contact> container, RowView rowView) {
    ...
  }

  List<Contact> result(Map<Long, Contact> container) {
    return new ArrayList<>(container.values());
  }
}
```

```java
class OptionalContactPhoneReducer
    implements RowReducer<Map<Long, Contact>, Optional<Contact>> {
  Map<Long, Contact> container() {
    return new LinkedHashMap<>();
  }

  void accumulate(Map<Long, Contact> container, RowView rowView) {
    ...
  }

  Optional<Contact> result(Map<Long, Contact> container) {
    return container.values().stream().findFirst();
  }
}
```

This felt wasteful to me, especially considering we already have `JdbiCollectors` in place to facilitate collecting elements into arbitrary container types.

I considered putting the `Stream<T>` restriction only in SQL Object, but that would have left us in a situation where you have to implement a `RowReducer` differently in core than in sqlobject.

Ultimately I concluded that it was better to have it return a stream, for consistency and possibility for reuse between core and sqlobject.

This is why I added `ResultBearing.reduceRows(Collector<RowView, ?, R>)`, as a compromise so users could have it either way.

@stevenschlansker Does this sound reasonable?

Reply:
Yeah, that's a real bummer.  I agree with your reasoning.  I'm still sad that the example already shows an example of the problem though -- this pattern essentially guarantees unnecessary copies.